crates/holochain_state/src/graft_records.rs

```diff
@@ -1,186 +1,113 @@
-use holochain_state::source_chain::SourceChain;
-use holochain_types::prelude::ChainItem;
+use holochain_chc::{ChcError, ChcImpl};
 
-use super::*;
-
-#[cfg_attr(feature = "instrument", tracing::instrument(skip_all))]
-pub(crate) async fn graft_records_onto_source_chain(
-    conductor: ConductorHandle,
-    cell_id: CellId,
-    validate: bool,
-    records: Vec<Record>,
-) -> ConductorApiResult<()> {
-    // Require that the cell is installed.
-    if let err @ Err(ConductorError::CellMissing(_)) = conductor.cell_by_id(&cell_id).await {
-        let _ = err?;
-    }
-
-    // Get or create the space for this cell.
-    let space = conductor.get_or_create_space(cell_id.dna_hash())?;
-
-    let chc = None;
-    let network = conductor
-        .holochain_p2p()
-        .to_dna(cell_id.dna_hash().clone(), chc);
-
-    let source_chain: SourceChain = space
-        .source_chain(conductor.keystore().clone(), cell_id.agent_pubkey().clone())
-        .await?;
-
-    let existing = source_chain
-        .query(ChainQueryFilter::new().descending())
-        .await?
-        .into_iter()
-        .map(|r| r.signed_action)
-        .collect::<Vec<SignedActionHashed>>();
+use crate::prelude::*;
+use crate::source_chain::SourceChain;
 
-    let graft = ChainGraft::new(existing, records).rebalance();
-    let chain_top = graft.existing_chain_top();
+use super::*;
 
-    if validate {
-        validate_records(conductor.clone(), &cell_id, &chain_top, graft.incoming()).await?;
+impl SourceChain {
+    /// Attempt to add the given records to the CHC.
+    /// If the CHC is already in sync with the local state, it will accept the new records and return `Ok`.
+    /// If not, the local state will instead be updated, the new records will not be accepted, and an `Err` will be returned.
+    //
+    // TODO: add Scratch awareness to this, so that [`ChainTopOrdering::Relaxed`] can be respected.
+    pub async fn sync_records(
```

Comment:
This is only dealing with one possible way round that this goes wrong right? The `InvalidChain` error should tell us that we're behind the CHC but what about when we're ahead? That's going to be relevant with the changes to countersigning and is a problem in general. If we're ahead and pushing wouldn't fork then shouldn't we push records that are held locally that aren't on the CHC?

That's also relevant for somebody who turns on the CHC after they've started authoring data I think. They won't be able to turn it on otherwise, which makes it a very all-or-nothing feature that you have to know you want to opt into up front.

Reply:
If we're ahead of the CHC and there is no forking, then we can just do `add_records` to advance the CHC. This is how somebody would get started with the CHC using an existing local chain. (There is no UX for this but it could be provided.)

The CHC is currently designed so that no local change happens without first updating the CHC.
If we allow local chains to commit and publish data before updating the CHC, there is the chance that other local chains can also concurrently commit changes and cause a fork. 
The entire point of the CHC is to prevent this possibility.

You're right that countersigning throws a wrench in the works. I think the thing to do here is actually to allow a client to *lock* the CHC, just as they lock their own chain, during countersigning. Because really, that lock should apply to all instances of your chain out there, not just the local one, and it's the job of the CHC to coordinate that.

Countersigning is an exceptional case where we commit things to the chain which can be eventually removed. The only reason that works is because we lock the chain, so that from an outside perspective, if the chain got rolled back, there was never a time when the chain was unlocked where the rolled back records were available. So it makes sense to me to extend that lock to the CHC, rather than simply deferring updating the CHC until after the local lock has been completed (which opens the door to conflicts from other chain instances). What do you think?

Reply:
@ThetaSinner to clarify, I'm proposing that during a countersigning session, we don't push anything to the CHC, and only do a CHC push after the chain (and CHC) is unlocked.

That allows one to freely participate in the session they're in without being restricted by the activity of the other chains managed by the CHC, and without having to communicate to the other chains that they're in a countersigning session. And come to think of it, it's really gnarly to consider pushing countersigned entries to the CHC during a session, because that implies that other chains can pick them up and also be participants in the same session. I think we really want to prevent that, and limit CS participation to a single node, and then communicate the result to the CHC after the fact.

If that's still not clear, then maybe we should both lay out our proposed order of operations. Mine is:

- before even accepting CS session, lock the CHC
- now I know that my CHC chain is not going to advance during the session, and I can accept and lock my local chain
- do all the normal CS stuff, and unlock my local chain at the end
- if at the end I throw out the session and rollback, I simply unlock the CHC
- if the session was successful, I push the countersigned entries to the CHC while unlocking (a single atomic operation)

Reply:
Another way to say it is that, if you're using a CHC, and you need to lock your chain, it probably makes sense to lock every other chain that may be out there as well. The point of locking your chain is to prevent other activity from altering the chain during the operation for which you locked, and that should probably extend to preventing any activity that would alter *any* instance of your chain.

Reply:
Following the discussion in MM, I'm agreeing that we don't need this and we can look at locking the CHC to support the way that countersigning is now using the CHC

Reply:
Could you double check that I merged in the countersigning work sensibly, in SourceChain::flush ?